File: problems/75-sort-colors/solution.ts
function sortColors(nums: number[]): number[] {
  if (nums.length <= 1) return nums;

  let left = 0,
    mid = 0,
    right = nums.length - 1;

  while (mid <= right) {
    if (nums[mid] === 0) {
      if (left !== mid) {
        nums[left] ^= nums[mid];
        nums[mid] ^= nums[left];
        nums[left] ^= nums[mid];
      }
      left++;
      mid++;
    } else if (nums[mid] === 2) {
      if (right !== mid) {
        nums[right] ^= nums[mid];
        nums[mid] ^= nums[right];
        nums[right] ^= nums[mid];
      }
      right--;
    } else mid++;
  }

  return nums;
}

export { sortColors };

File: problems/75-sort-colors/solution.go
package main

func sortColors(nums []int) []int {
	if cap(nums) <= 1 {
		return nums
	}

	left, right := 0, len(nums)-1

	for mid := 0; mid <= right; {
		switch nums[mid] {
		case 0:
			if left != mid {
				nums[left] ^= nums[mid]
				nums[mid] ^= nums[left]
				nums[left] ^= nums[mid]
			}
			left++
			mid++
		case 2:
			if right != mid {
				nums[right] ^= nums[mid]
				nums[mid] ^= nums[right]
				nums[right] ^= nums[mid]
			}
			right--
		default:
			mid++
		}
	}

	return nums
}
